Add sales channel route and configuration for sw-domain-hash

JoshuaBehrens opened the pull request:
### 1. Why is this change necessary?
File upload/DNS settings are less common to be easily available for persons that setup shops and handle shopware accounts.

### 2. What does this change do, exactly?
Add a configuration field for sales channels that are served under /sw-domain-hash.html

### 3. Describe each step to reproduce the issue or behaviour.
1. Have a coworker setting up shops
2. Coworkers request me to upload a file
3. Me being annoyed as everything is pipeline controlled
4. Rather choose the DNS way
5. Change a DNS setting that might need some time to be populated across servers
6. 🙄

### 4. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### 5. Additionals

This should be added as well to the account.shopware.com verification process so people are aware that this option is available without this high technical obstacle.

janbuecker commented:
Thanks @kayyyy 

We decided to close this PR because of time constraints and limited overall benefit to the product. If this change is still important to you, please create a new issue on https://issues.shopware.com so people can up-vote this feature.

JoshuaBehrens committed:
Add sales channel route and configuration for sw-domain-hash

hlohaus commented:
Why on sales channel? It's only one possible. Maybe a configuration are better?

JoshuaBehrens commented:
> Why on sales channel?

AFAIK this code is per domain. And domains are on sales channels. But I think @shopwareBot can clear things up a bit.

shyim commented:
Nope. It's currently handled by an option in the store.xml. This option is for the entire installation.

JoshuaBehrens commented:
> Nope. It's currently handled by an option in the store.xml. This option is for the entire installation.

@shyim what?

shyim commented:
https://github.com/shopware/platform/blob/6.1/src/Core/System/Resources/config/store.xml#L10
That's the license host for the verification. And this is configurable for the entire installation only

JoshuaBehrens commented:
So I just add a configuration there for sw-domain-hash as well?

shyim commented:
Just wait for tomorrow for the decision from shopwareBot :)

JoshuaBehrens commented:
@shopwareBot @OliverSkroblin so what is the decision? I still support the idea to have it on a sales channel as a single installation can serve multiple clients at once.

JoshuaBehrens commented:
I just rebased it to keep things fresh :)

JoshuaBehrens commented:
Fresh like a fridge

pweyck commented:
After some discussion with shopwareBot, we came to the conclusion, that we can add this. But it should be system wide like https://github.com/shopware/platform/blob/6.1/src/Core/System/Resources/config/store.xml#L10. We also need to ensure that the controller is accessible even if the maintenance mode is enabled. Maybe also add a tooltip that the file on the web server takes precedence.

pweyck commented:
Hi @JoshuaBehrens, do you still want to work this?

We would accept it if
- it's stored system wide in system_config (sales_channel_id === NULL). See: https://github.com/shopware/platform/blob/6.1/src/Core/System/Resources/config/store.xml#L10
- works in maintenance mode

JoshuaBehrens commented:
Hi @pweyck ,

thank you for the reminder. A config can also show the same content, so it looks easy to fix. I put it on my list.

JoshuaBehrens committed:
Add route attribute to whitelist routes for maintenance

JoshuaBehrens commented:
Refactored regarding the comments. I took the liberty and introduced a request attribute/route default to make it easier and new default way to whitelist a route for maintenance access.

JoshuaBehrens commented:
I added docs how to whiteliste a route for maintenance mode

mitelg commented on src/Storefront/Framework/Routing/MaintenanceModeResolver.php:
```diff
@@ -44,12 +44,17 @@ private function isSalesChannelRequest(?Request $master): bool
             return false;
         }
 
+        // @deprecated tag:v6.5.0 - Use defaults={"allow_maintenance"=true} in Route definition instead
```
6.4.0 should be okay

JoshuaBehrens replied:
Nice

shyim commented on src/Storefront/Controller/VerificationHashController.php:
```diff
@@ -0,0 +1,48 @@
+    public function load(SalesChannelContext $context): Response
+    {
+        $salesChannelId = $context->getSalesChannel()->getId();
+        $verificationHash = $this->systemConfigService->get('core.store.verificationHash', $salesChannelId);
```
```suggestion
        $verificationHash = $this->systemConfigService->get('core.store.verificationHash');
```
Only the instance has to be verified

JoshuaBehrens replied:
But I can change it per saleschannel, am I right?

shyim replied:
The shop domain in account is based on store license host. This is per installation and not per sales channel

JoshuaBehrens replied:
Oh, I checked the admin interface. There is no sales channel switch. Nice

shyim commented on src/Storefront/Controller/VerificationHashController.php:
```diff
@@ -0,0 +1,48 @@
+use Symfony\Component\Routing\Annotation\Route;
+
+/**
+ * @RouteScope(scopes={"storefront"})
```
This assumes that the Store License domain must be an sales channel. The licenseHost can currently be anything.
https://github.com/shopware/platform/pull/572/files#diff-bfa3c20d9d35a22675e5b7e5adc13bdfR9

So the controller is not accessible then

JoshuaBehrens replied:
Which scope do you suggest instead? IIRC there is an exception when no scope is given.

pweyck replied:
We should change it to the api scope, because it's part of the core. 
```suggestion
 * @RouteScope(scopes={"api"})
```

JoshuaBehrens replied:
When I change the scope I have to add the attribute that I don't need credentials for authentication, right?

pweyck replied:
You're too fast :) See my newest comment

JoshuaBehrens replied:
You're too slow! 
![grafik](https://user-images.githubusercontent.com/1133593/90012716-38911180-dca4-11ea-918b-7c59ebc67f9f.png)
I'll wait next time :D I see where this is going

shyim commented on src/Storefront/Exception/VerificationHashNotConfiguredException.php:
```diff
@@ -0,0 +1,28 @@
+
+class VerificationHashNotConfiguredException extends ShopwareHttpException
+{
+    public function __construct(string $salesChannelId, ?\Throwable $e)
```
```suggestion
    public function __construct(string $salesChannelId, ?\Throwable $e = null)
```

pweyck commented on src/Storefront/Controller/VerificationHashController.php:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    /**
+     * @Route("/sw-domain-hash.html", name="frontend.verification-hash.load", methods={"GET"}, defaults={"allow_maintenance"=true})
```
```suggestion
     * @Route("/sw-domain-hash.html", name="api.verification-hash.load", methods={"GET"}, defaults={"auth_required"=false})
```

pweyck commented on src/Storefront/Exception/VerificationHashNotConfiguredException.php:
```diff
@@ -0,0 +1,28 @@
+
+class VerificationHashNotConfiguredException extends ShopwareHttpException
+{
+    public function __construct(string $salesChannelId, ?\Throwable $e = null)
```
This should be sales channel independent.

pweyck reviewed (changes requested):
You also need to add `/sw-domain-hash.html` to https://github.com/shopware/platform/blob/feba2eaced67b06547deffb857cf3d52c90e85fa/src/Core/Framework/Routing/ApiRouteScope.php#L18

JoshuaBehrens commented:
Hot as a fresh cup cake. Rebase is done as well

bneumann97 commented:
Hey @JoshuaBehrens, thanks for your contribution 💙 Our issuetracker import tool doesn't like your description again, therefore I have created a ticket for you manually https://issues.shopware.com/issues/NEXT-12003

bneumann97 commented:
Hey @JoshuaBehrens, thanks for your contribution 💙

JoshuaBehrens commented:
Thank you for merging :)
